cms/bundles/viewsets.py

```diff
@@ -0,0 +1,308 @@
+from django.http import HttpRequest
+from django.shortcuts import redirect
+from django.urls import reverse
+from django.utils import timezone
+from django.utils.html import format_html, format_html_join
+from django.utils.translation import gettext as _
+from wagtail.admin.ui.tables import Column, DateColumn, UpdatedAtColumn, UserColumn
+from wagtail.admin.views.generic import CreateView, EditView, IndexView, InspectView
+from wagtail.admin.views.generic.chooser import ChooseView
+from wagtail.admin.viewsets.chooser import ChooserViewSet
+from wagtail.admin.viewsets.model import ModelViewSet
+from wagtail.log_actions import log
+
+from .enums import BundleStatus
+from .models import Bundle
+from .notifications import notify_slack_of_publication_start, notify_slack_of_publish_end, notify_slack_of_status_change
+
+if TYPE_CHECKING:
+    from django.db.models.fields import Field
+    from django.http import HttpResponseBase
+    from django.utils.safestring import SafeString
+
+
+class BundleCreateView(CreateView):
+    """The Bundle create view class."""
+
+    def save_instance(self) -> Bundle:
+        """Automatically set the creating user on Bundle creation."""
+        instance: Bundle = super().save_instance()
+        instance.created_by = self.request.user
+        instance.save(update_fields=["created_by"])
+        return instance
+
+
+class BundleEditView(EditView):
+    """The Bundle edit view class."""
+
+    actions: ClassVar[list[str]] = ["edit", "save-and-approve", "publish"]
+    template_name = "bundles/wagtailadmin/edit.html"
+    has_content_changes: bool = False
+    start_time: float | None = None
+
+    def dispatch(self, request: HttpRequest, *args: Any, **kwargs: Any) -> "HttpResponseBase":
+        if (instance := self.get_object()) and instance.status == BundleStatus.RELEASED:
+            return redirect(self.index_url_name)
+
+        response: HttpResponseBase = super().dispatch(request, *args, **kwargs)
+        return response
+
+    def get_form_kwargs(self) -> dict:
+        kwargs: dict = super().get_form_kwargs()
+        if self.request.method == "POST":
+            data = self.request.POST.copy()
+            if "action-save-and-approve" in self.request.POST:
+                data["status"] = BundleStatus.APPROVED.value
+                data["approved_at"] = timezone.now()
+                data["approved_by"] = self.request.user
+                kwargs["data"] = data
+            elif "action-publish" in self.request.POST:
+                data["status"] = BundleStatus.RELEASED.value
+                kwargs["data"] = data
+        return kwargs
+
+    def save_instance(self) -> Bundle:
+        instance: Bundle = self.form.save()
+        self.has_content_changes = self.form.has_changed()
+
+        if self.has_content_changes:
+            log(action="wagtail.edit", instance=instance, content_changed=True, data={"fields": self.form.changed_data})
+
+            if "status" in self.form.changed_data:
```

Comment:
Minor, can we inverse the logic to exit early. Helps reduce indentation and improves readability. This is a pattern I would like to adopt across the project where appropriate.

```py
if not self.has_content_changes:
	return

...
```